### DISCUSSION TITLE
Get points coordinates from triangulation mesh

### DISCUSSION THREAD

**Discussion by elifoo** [pyvista/pyvista#1465] [Q&A] [answered]
Hello Pyvista team,

I obtained a triangular mesh with  pv.PolyData(xyz).delaunay_3d(), where xyz is a 'numpy.ndarray' with shape (N,3).
I'd like get information about the points coordinates that built the mesh, i.e. the triplets of the vertices of each triangle (n,3,3). Can you help me?
Thanks so much!

**Comment by akaszynski** [comment, marked as answer]
Hello,

You can grab the points of the mesh with
```pyvista
import pyvista
...
mesh = pv.PolyData(xyz).delaunay_3d()
mesh.points
```

**Reply by elifoo** [reply]
Thank you,
but maybe I didn't explain myself very well.
mesh.points returns the original points (xyz in my case with shape (N, 3)). I would like, the coordinates of the single vertices of the triangles of the mesh ... so I should get an array of (n,3,3) with n> N (due to vertices in common with other triangles)

**Reply by akaszynski** [reply]
Here's how you would get an `N x 3 x 3` array of points (points per triangle for all triangles).  It's not an efficient way of storing the points, but certain libraries like to store points this way.

```py
First create the mesh.  Verify it's all triangles

>>> mesh = pyvista.Sphere()
>>> mesh.is_all_triangles()
True

Extract just the face indices.  Realize that VTK adds padding by default (the number of points per face).

>>> faces = mesh.faces.reshape((-1, 4))  # can do this since it's all triangles
>>> faces
array([[  3,   2,  30,   0],
       [  3,  30,  58,   0],
       [  3,  58,  86,   0],
       ...,
       [  3, 839,  28,  27],
       [  3, 840, 841,  29],
       [  3, 840,  29,  28]])

# we just want the face indices

>>> faces = faces[:, 1:]
>>> faces
array([[  2,  30,   0],
       [ 30,  58,   0],
       [ 58,  86,   0],
       ...,
       [839,  28,  27],
       [840, 841,  29],
       [840,  29,  28]])

Now we can construct our Nx3x3 point array

>>> arr = mesh.points[faces]
>>> arr.shape
(1680, 3, 3)
```

**Reply by elifoo** [reply]
Great!
Now It works very well!
I just added mesh_ok = mesh.extract_surface() , before extracting the faces.

Thanks a lot!
